File: iFinance/AppStore.swift
import SwiftUI

@Observable
class AppStore {
    var transactions: [Transaction] = []
    var streakDays: Int = 0
    var lastRecordDate: Date? = nil
    var longestStreak: Int = 0
    var xp: Int = 0
    var monthlyBudget: Double = 0 {
        didSet { save() }
    }
    var unlockedAchievements: Set<String> = []

    // Reactive state for UI
    var showCelebration = false
    var lastAddedTransaction: Transaction? = nil
    var newlyUnlockedAchievement: AchievementDef? = nil

    var level: Int { max(1, xp / 100 + 1) }
    var xpInCurrentLevel: Int { xp % 100 }

    var totalBalance: Double {
        transactions.reduce(0) { $0 + ($1.type == .income ? $1.amount : -$1.amount) }
    }

    var totalIncome: Double {
        transactions.filter { $0.type == .income }.reduce(0) { $0 + $1.amount }
    }

    var totalExpense: Double {
        transactions.filter { $0.type == .expense }.reduce(0) { $0 + $1.amount }
    }

    var thisMonthTransactions: [Transaction] {
        let calendar = Calendar.current
        let now = Date()
        return transactions.filter {
            calendar.isDate($0.date, equalTo: now, toGranularity: .month)
        }
    }

    var thisMonthExpense: Double {
        thisMonthTransactions.filter { $0.type == .expense }.reduce(0) { $0 + $1.amount }
    }

    var budgetRemaining: Double {
        monthlyBudget - thisMonthExpense
    }

    var budgetProgress: Double {
        guard monthlyBudget > 0 else { return 0 }
        return min(thisMonthExpense / monthlyBudget, 1.0)
    }

    var expenseByCategory: [(category: Category, amount: Double)] {
        let expenses = thisMonthTransactions.filter { $0.type == .expense }
        var dict: [String: Double] = [:]
        for t in expenses {
            dict[t.category, default: 0] += t.amount
        }
        return dict.compactMap { key, value in
            guard let cat = CATEGORIES.first(where: { $0.id == key }) else { return nil }
            return (category: cat, amount: value)
        }.sorted { $0.amount > $1.amount }
    }

    init() {
        load()
        validateStreak()
    }

    func addTransaction(_ t: Transaction) {
        transactions.append(t)
        xp += 10
        lastAddedTransaction = t
        updateStreak()
        let newBadge = checkAchievements()
        save()

        showCelebration = true
        if let badge = newBadge {
            newlyUnlockedAchievement = badge
        }
    }

    // MARK: - Streak

    private func updateStreak() {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())

        if let last = lastRecordDate {
            let lastDay = calendar.startOfDay(for: last)
            if lastDay == today { return }
            let diff = calendar.dateComponents([.day], from: lastDay, to: today).day ?? 0
            if diff == 1 {
                streakDays += 1
            } else {
                streakDays = 1
            }
        } else {
            streakDays = 1
        }
        lastRecordDate = today
        longestStreak = max(longestStreak, streakDays)
    }

    /// Validate streak on app launch — reset if more than 1 day has passed since last record
    private func validateStreak() {
        guard let last = lastRecordDate else { return }
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        let lastDay = calendar.startOfDay(for: last)
        let diff = calendar.dateComponents([.day], from: lastDay, to: today).day ?? 0
        // If more than 1 day gap, streak is broken
        if diff > 1 {
            streakDays = 0
            save()
        }
    }

    // MARK: - Achievements

    @discardableResult
    private func checkAchievements() -> AchievementDef? {
        var newlyUnlocked: AchievementDef? = nil

        func tryUnlock(_ id: String, condition: Bool) {
            if condition && !unlockedAchievements.contains(id) {
                unlockedAchievements.insert(id)
                if let def = ALL_ACHIEVEMENTS.first(where: { $0.id == id }) {
                    newlyUnlocked = def
                }
            }
        }

        let count = transactions.count
        let balance = totalBalance

        tryUnlock("first_record", condition: count >= 1)
        tryUnlock("records_10", condition: count >= 10)
        tryUnlock("records_30", condition: count >= 30)
        tryUnlock("records_100", condition: count >= 100)
        tryUnlock("streak_3", condition: streakDays >= 3)
        tryUnlock("streak_7", condition: streakDays >= 7)
        tryUnlock("streak_30", condition: streakDays >= 30)
        tryUnlock("save_1000", condition: balance >= 1000)
        tryUnlock("save_10000", condition: balance >= 10000)
        tryUnlock("level_5", condition: level >= 5)
        tryUnlock("level_10", condition: level >= 10)

        return newlyUnlocked
    }

    // MARK: - Persistence

    func save() {
        let encoder = JSONEncoder()
        if let data = try? encoder.encode(transactions) {
            UserDefaults.standard.set(data, forKey: "transactions")
        }
        UserDefaults.standard.set(streakDays, forKey: "streakDays")
        if let date = lastRecordDate {
            UserDefaults.standard.set(date.timeIntervalSince1970, forKey: "lastRecordDate")
        }
        UserDefaults.standard.set(longestStreak, forKey: "longestStreak")
        UserDefaults.standard.set(xp, forKey: "xp")
        UserDefaults.standard.set(monthlyBudget, forKey: "monthlyBudget")
        if let data = try? encoder.encode(Array(unlockedAchievements)) {
            UserDefaults.standard.set(data, forKey: "unlockedAchievements")
        }
    }

    private func load() {
        let decoder = JSONDecoder()
        if let data = UserDefaults.standard.data(forKey: "transactions"),
           let decoded = try? decoder.decode([Transaction].self, from: data) {
            transactions = decoded
        }
        streakDays = UserDefaults.standard.integer(forKey: "streakDays")
        let ts = UserDefaults.standard.double(forKey: "lastRecordDate")
        if ts > 0 { lastRecordDate = Date(timeIntervalSince1970: ts) }
        longestStreak = UserDefaults.standard.integer(forKey: "longestStreak")
        xp = UserDefaults.standard.integer(forKey: "xp")
        monthlyBudget = UserDefaults.standard.double(forKey: "monthlyBudget")
        if let data = UserDefaults.standard.data(forKey: "unlockedAchievements"),
           let decoded = try? decoder.decode([String].self, from: data) {
            unlockedAchievements = Set(decoded)
        }
    }
}

File: lib/stores/app_store.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/transaction.dart';
import '../models/category.dart';
import '../models/achievement.dart';

class AppStore extends ChangeNotifier {
  List<Transaction> transactions = [];
  int streakDays = 0;
  DateTime? lastRecordDate;
  int longestStreak = 0;
  int xp = 0;
  double monthlyBudget = 0;
  Set<String> unlockedAchievements = {};

  // Reactive UI state
  bool showCelebration = false;
  Transaction? lastAddedTransaction;
  AchievementDef? newlyUnlockedAchievement;

  int get level => (xp ~/ 100) + 1;
  int get xpInCurrentLevel => xp % 100;

  double get totalBalance => transactions.fold(
      0.0, (sum, t) => sum + (t.type == TransactionType.income ? t.amount : -t.amount));

  double get totalIncome => transactions
      .where((t) => t.type == TransactionType.income)
      .fold(0.0, (sum, t) => sum + t.amount);

  double get totalExpense => transactions
      .where((t) => t.type == TransactionType.expense)
      .fold(0.0, (sum, t) => sum + t.amount);

  List<Transaction> get thisMonthTransactions {
    final now = DateTime.now();
    return transactions
        .where((t) => t.date.year == now.year && t.date.month == now.month)
        .toList();
  }

  double get thisMonthExpense => thisMonthTransactions
      .where((t) => t.type == TransactionType.expense)
      .fold(0.0, (sum, t) => sum + t.amount);

  double get budgetRemaining => monthlyBudget - thisMonthExpense;

  double get budgetProgress {
    if (monthlyBudget <= 0) return 0;
    return (thisMonthExpense / monthlyBudget).clamp(0.0, 1.0);
  }

  List<MapEntry<CategoryDef, double>> get expenseByCategory {
    final expenses =
        thisMonthTransactions.where((t) => t.type == TransactionType.expense);
    final dict = <String, double>{};
    for (final t in expenses) {
      dict[t.category] = (dict[t.category] ?? 0) + t.amount;
    }
    final result = dict.entries
        .map((e) => MapEntry(getCategoryById(e.key), e.value))
        .toList();
    result.sort((a, b) => b.value.compareTo(a.value));
    return result;
  }

  AppStore() {
    _load().then((_) {
      _validateStreak();
      notifyListeners();
    });
  }

  void addTransaction(Transaction t) {
    transactions.add(t);
    xp += 10;
    lastAddedTransaction = t;
    _updateStreak();
    final newBadge = _checkAchievements();
    _save();

    showCelebration = true;
    if (newBadge != null) {
      newlyUnlockedAchievement = newBadge;
    }
    notifyListeners();
  }

  void setMonthlyBudget(double value) {
    monthlyBudget = value;
    _save();
    notifyListeners();
  }

  void dismissCelebration() {
    showCelebration = false;
    notifyListeners();
  }

  void dismissAchievementToast() {
    newlyUnlockedAchievement = null;
    notifyListeners();
  }

  // -- Streak --

  void _updateStreak() {
    final today = DateUtils.dateOnly(DateTime.now());
    if (lastRecordDate != null) {
      final lastDay = DateUtils.dateOnly(lastRecordDate!);
      if (lastDay == today) return;
      final diff = today.difference(lastDay).inDays;
      if (diff == 1) {
        streakDays += 1;
      } else {
        streakDays = 1;
      }
    } else {
      streakDays = 1;
    }
    lastRecordDate = today;
    if (streakDays > longestStreak) longestStreak = streakDays;
  }

  void _validateStreak() {
    if (lastRecordDate == null) return;
    final today = DateUtils.dateOnly(DateTime.now());
    final lastDay = DateUtils.dateOnly(lastRecordDate!);
    final diff = today.difference(lastDay).inDays;
    if (diff > 1) {
      streakDays = 0;
      _save();
    }
  }

  // -- Achievements --

  AchievementDef? _checkAchievements() {
    AchievementDef? newlyUnlocked;

    void tryUnlock(String id, bool condition) {
      if (condition && !unlockedAchievements.contains(id)) {
        unlockedAchievements.add(id);
        final def = kAllAchievements.where((a) => a.id == id).firstOrNull;
        if (def != null) newlyUnlocked = def;
      }
    }

    final count = transactions.length;
    final balance = totalBalance;

    tryUnlock('first_record', count >= 1);
    tryUnlock('records_10', count >= 10);
    tryUnlock('records_30', count >= 30);
    tryUnlock('records_100', count >= 100);
    tryUnlock('streak_3', streakDays >= 3);
    tryUnlock('streak_7', streakDays >= 7);
    tryUnlock('streak_30', streakDays >= 30);
    tryUnlock('save_1000', balance >= 1000);
    tryUnlock('save_10000', balance >= 10000);
    tryUnlock('level_5', level >= 5);
    tryUnlock('level_10', level >= 10);

    return newlyUnlocked;
  }

  // -- Persistence --

  Future<void> _save() async {
    final prefs = await SharedPreferences.getInstance();
    final txJson = transactions.map((t) => t.toJson()).toList();
    await prefs.setString('transactions', jsonEncode(txJson));
    await prefs.setInt('streakDays', streakDays);
    if (lastRecordDate != null) {
      await prefs.setInt(
          'lastRecordDate', lastRecordDate!.millisecondsSinceEpoch);
    }
    await prefs.setInt('longestStreak', longestStreak);
    await prefs.setInt('xp', xp);
    await prefs.setDouble('monthlyBudget', monthlyBudget);
    await prefs.setStringList(
        'unlockedAchievements', unlockedAchievements.toList());
  }

  Future<void> _load() async {
    final prefs = await SharedPreferences.getInstance();
    final txString = prefs.getString('transactions');
    if (txString != null) {
      final list = jsonDecode(txString) as List;
      transactions = list.map((e) => Transaction.fromJson(e)).toList();
    }
    streakDays = prefs.getInt('streakDays') ?? 0;
    final ts = prefs.getInt('lastRecordDate');
    if (ts != null) lastRecordDate = DateTime.fromMillisecondsSinceEpoch(ts);
    longestStreak = prefs.getInt('longestStreak') ?? 0;
    xp = prefs.getInt('xp') ?? 0;
    monthlyBudget = prefs.getDouble('monthlyBudget') ?? 0;
    unlockedAchievements =
        (prefs.getStringList('unlockedAchievements') ?? []).toSet();
  }
}
